csharp/extractor/Semmle.Extraction.CIL/Factories.cs

```diff
@@ -9,37 +11,78 @@ namespace Semmle.Extraction.CIL
     /// </summary>
     public partial class Context
     {
-        readonly Dictionary<Id, (Label, Id)> ids = new Dictionary<Id, (Label, Id)>();
+        readonly Dictionary<object, Label> ids = new Dictionary<object, Label>();
 
-        public T Populate<T>(T e) where T : ILabelledEntity
+        public T Populate<T>(T e) where T : IExtractedEntity
         {
-            Id id = e.ShortId;
+            if(e.Label.Valid)
+            {
+                return e;   // Already populated
+            }
 
-            if (ids.TryGetValue(id, out var existing))
+            if (ids.TryGetValue(e, out var existing))
             {
                 // It exists already
-                e.Label = existing.Item1;
-                e.ShortId = existing.Item2;   // Reuse ID for efficiency
+                e.Label = existing;
             }
             else
             {
-                cx.DefineLabel(e);
-                ids.Add(id, (e.Label, id));
+                e.Label = cx.GetNewLabel();
+                cx.DefineLabel(e, cx.TrapWriter.Writer);
+                ids.Add(e, e.Label);
                 cx.PopulateLater(() =>
                 {
                     foreach (var c in e.Contents)
                         c.Extract(this);
                 });
+#if DEBUG_LABELS
+                using (var writer = new StringWriter())
+                {
+                    e.WriteId(writer);
+                    var id = writer.ToString();
+
+                    if (debugLabels.TryGetValue(id, out IExtractedEntity previousEntity))
+                    {
+                        cx.Extractor.Message(new Message("Duplicate trap ID", id, null, severity: Util.Logging.Severity.Warning));
+                    }
+                    else
+                    {
+                        debugLabels.Add(id, e);
+                    }
+                }
+#endif
             }
             return e;
         }
 
+#if DEBUG_LABELS
+        private readonly Dictionary<string, IExtractedEntity> debugLabels = new Dictionary<string, IExtractedEntity>();
+#endif
+
         public IExtractedEntity Create(Handle h)
         {
             var entity = CreateGeneric(defaultGenericContext, h);
             return entity;
         }
 
+        // Lazily cache primitive types.
+        private readonly PrimitiveType[] primitiveTypes = new PrimitiveType[(int)PrimitiveTypeCode.Object + 1];
```

Comment:
How about `new PrimitiveType[Enum.GetValues(typeof(PrimitiveTypeCode)).Cast<int>().Max() + 1]`?

Reply:
Ooh funky. I think I'll keep the simpler version as I don't think this is going to change - and if it ever does then many other things will also break anyway.

Reply:
OK.